### PR TITLE
Don't interpret type path as part of visibility.

### LINKED ISSUE #5902: Syntax error on public tuples referencing super
These two snippets show syntax errors when they aren't
```rust
pub struct A;

mod module {
    pub struct B(pub (super::A,));
}
```
```rust
pub struct A;

mod module {
    pub struct B(pub (super::A));
}
```

image

Hovering over the errors gives `Syntax Error: expected R_PAREN` for the first squiggle and both `Syntax Error: expected SEMICOLON` and two separate `Syntax Error: expected an item` for the second

image

### PR DESCRIPTION
**Pull request by ArifRoktim**
This closes #5902.
I only check that the next token isn't equal to `T![:]`, instead of the next two not being equal to `T![::]`. Is that ok?

### CODE REVIEW

File: crates/parser/src/grammar.rs [hunk 1]
```diff
@@ -185,9 +185,14 @@ fn opt_visibility(p: &mut Parser) -> bool {
                     // pub(self) struct S;
                     // pub(self) struct S;
                     T![crate] | T![self] | T![super] => {
-                        p.bump_any();
-                        p.bump_any();
-                        p.expect(T![')']);
+                        // test pub_parens_typepath
+                        // struct B(pub (super::A));
+                        // struct B(pub (crate::A,));
+                        if p.nth(2) != T![:] {
```
Comment (matklad) [inline crates/parser/src/grammar.rs:191]: Let's move this condition to the match arm guard.

Bonus points: fix **Move condition to match guard** assists such that not only this works:

image


But this as well:

image

**Comment by matklad** [comment]
>Is that ok?

Yup!

File: crates/parser/src/grammar.rs (hunk 1 above)
Comment (ArifRoktim) [reply crates/parser/src/grammar.rs]: Don't merge yet.
While I was eating my own dog food,
dog_food1
dog_food2
 I realized that comments sometimes get eaten:
```rust
fn main() {
    match 92 {
        x => {
            // This comment gets eaten
            if x > 10 {
                92;
                // This comment survives
                false
            }
        }
        _ => true
    }
}
```
Will fix tomorrow.

Comment (ArifRoktim) [reply crates/parser/src/grammar.rs]: How should comments in blocks be handled? Does the behavior of these tests make sense?

<details>
<summary>Tests</summary>

```rust
    fn move_arm_cond_with_comments_to_match_guard_works() {
        check_assist(
            move_arm_cond_to_match_guard,
            r#"
fn main() {
    match 92 {
        // Comment A
        x => if x > 10 {<|>
            // Comment B
            false
        },
        _ => true
    }
}
"#,
            r#"
fn main() {
    match 92 {
        // Comment A
        // Comment B
        x if x > 10 => false,
        _ => true
    }
}
"#,
        );
    }

    #[test]
    fn move_arm_cond_with_comments_in_block_to_match_guard_works() {
        check_assist(
            move_arm_cond_to_match_guard,
            r#"
fn main() {
    match 92 {
        x => {
            // Comment A
            if x > 10 {<|>
                // Comment B
                false
            }
        },
        _ => true
    }
}
"#,
            r#"
fn main() {
    match 92 {
        // Comment A
        // Comment B
        x if x > 10 => false,
        _ => true
    }
}
"#,
        );
    }
```
</details>

Or should those blocks not be "unwrapped" if there are comments?

Comment (matklad) [reply crates/parser/src/grammar.rs]: >How should comments in blocks be handled? Does the behavior of these tests make sense?

I think *somehow* is the right answer at this stage, we need more practical experience to figure out the best practicies here.

**Comment by matklad** [comment]
This looks good to go, so marking as ready and

bors r+

:)

Thanks @ArifRoktim , feel free to send a follow up PR if you want anything changed!

**Pull request merged**